tfhe/src/integer/server_key/radix_parallel/scalar_comparison.rs

```diff
@@ -0,0 +1,785 @@
+                    }
+                })
+                .collect::<Vec<_>>();
+        }
+
+        block_comparisons
+            .into_iter()
+            .next()
+            .expect("one block was expected")
+    }
+
+    /// This takes a Vec of shortint blocks, where each block is
+    /// either 0 or 1.
+    ///
+    /// It return a shortint block encrypting 1 if at least input blocks is 1
+    /// otherwise the block encrypts 0 (all blocks encrypts 0)
+    ///
+    /// if the vec is empty, a trivial 1 is returned
+    pub(crate) fn is_at_last_one_comparisons_block_true<PBSOrder>(
+        &self,
+        mut block_comparisons: Vec<CiphertextBase<PBSOrder>>,
+    ) -> CiphertextBase<PBSOrder>
+    where
+        PBSOrder: PBSOrderMarker,
+    {
+        if block_comparisons.is_empty() {
+            return self.key.create_trivial(1);
+        }
+
+        let message_modulus = self.key.message_modulus.0;
+        let carry_modulus = self.key.carry_modulus.0;
+        let total_modulus = message_modulus * carry_modulus;
+        let max_value = total_modulus - 1;
+
+        let is_not_zero = self
+            .key
+            .generate_accumulator(|x| u64::from((x & max_value as u64) != 0));
+
+        while block_comparisons.len() > 1 {
+            block_comparisons = block_comparisons
+                .par_chunks(max_value)
+                .map(|blocks| {
+                    let mut sum = blocks[0].clone();
+                    for other_block in &blocks[1..] {
+                        self.key.unchecked_add_assign(&mut sum, other_block);
+                    }
+
+                    self.key.apply_lookup_table(&sum, &is_not_zero)
+                })
+                .collect::<Vec<_>>();
+        }
+
+        block_comparisons
+            .into_iter()
+            .next()
+            .expect("one block was expected")
+    }
+
+    /// This takes an input slice of blocks.
+    ///
+    /// Each block can encrypt any value as long as its < message_modulus.
+    ///
+    /// It will compare blocks with 0, for either equality or difference.
+    ///
+    /// This returns a Vec of block, where each block encrypts 1 or 0
+    /// depending of if all blocks matched with the comparison type with 0.
+    ///
+    /// E.g. For ZeroComparisonType::Equality, if all input blocks are zero
+    /// than all returned block will encrypt 1
+    ///
+    /// The returned Vec will have less block than the number of input blocks.
+    /// The returned blocks potentialy needs to be 'reduced' to one block
+    /// with eg are_all_comparisons_block_true.
+    ///
+    /// This function exists because sometimes it is faster to concatenate
+    /// multiple vec of 'boolean' shortint block before reducing them with
+    /// are_all_comparisons_block_true
+    pub(crate) fn compare_blocks_with_zero<PBSOrder>(
+        &self,
+        lhs: &[CiphertextBase<PBSOrder>],
+        comparison_type: ZeroComparisonType,
```

Comment:
This could be moved into the comparator module, its where the `ZeroComparisonType` (I put it there to avoir having to make the scalar_comparisons pub(crate), comparator.rs is already pub)

Reply:
maybe in a subsequent PR ? having a big PR is always a bit tricky to finish reviews properly, I would like to merge this as it's functional and a big piece of code, what do you think ?

Reply:
ok for me

Reply:
alright